### PR TITLE
don't always copy LambdaInfo when adding methods

### PR DESCRIPTION
**Pull request by JeffBezanson**
It's very rare that code is written such that the same LambdaInfo is used by multiple method definitions. Therefore we should try to copy a LambdaInfo only when needed.

It also wasn't really intuitive to me that `types` should be the sole extra argument to a function for copying a LambdaInfo, so I removed the argument. Same applies to sparam_syms being the only argument to `jl_new_lambda_info`, but I'll deal with that later.

### CODE REVIEW

File: src/alloc.c [hunk 1]
```diff
@@ -476,8 +476,6 @@ static jl_lambda_info_t *jl_clone_thunk(jl_lambda_info_t *linfo, jl_tupletype_t
     new_linfo->nargs = linfo->nargs;
     new_linfo->isva = linfo->isva;
     new_linfo->rettype = linfo->rettype;
-    new_linfo->def = NULL;
-    new_linfo->specTypes = types;
```
Comment (yuyichao) [inline src/alloc.c]: Should this be left in or at least add a comment so user won't forget to initialize this?

Comment (JeffBezanson) [inline src/alloc.c]: Well, this is a static function, plus `jl_new_lambda_info_uninit` also leaves this field set to NULL. Also, I don't want to see more uses of this function. IMO, the public API of this subsystem is basically method_table_insert and get_specialization.

File: src/alloc.c [hunk 2]
```diff
@@ -563,19 +562,25 @@ jl_method_t *jl_new_method(jl_lambda_info_t *definition, jl_sym_t *name, jl_tupl
...
-        jl_cellset(stmts, i, jl_resolve_globals(jl_cellref(stmts, i), definition));
+    if (!reused) {
+        jl_array_t *stmts = definition->code;
+        int i, l;
+        for(i = 0, l = jl_array_len(stmts); i < l; i++) {
+            jl_cellset(stmts, i, jl_resolve_globals(jl_cellref(stmts, i), definition));
+        }
+        jl_method_init_properties(m);
     }
-    jl_method_init_properties(m);
```
Comment (vtjnash) [inline src/alloc.c]: I think we can copy these properties from `definition->def` if reused?

Comment (JeffBezanson) [inline src/alloc.c]: Yes good point.

**Pull request merged**